Filter notifications targets

AdrienClairembault opened the pull request:
## Changes summary

1. Add `FilterableInterface` interface to allow any `CommonDBTM` to define search engine based filters
2. Add `Item_Filter` class to store and interact with filters
3. Support filters for notifications targets by implementing `Item_Filter` and `Filterable` logic to Notifications
4. Add `hide_controls` parameter to the search engine
5. Add `extra_actions_templates` parameter to the search engine
6. Add `push_history` parameter to the search engine
7. Fix display parameters being lost on search submit
8. ~Fix js toasts inconsistencies~
9. Fix metacriteria search
10. ~Fix `getItemTypeForTable()` namespace handling~
11. ~Add `CommonDBTM::getDefaultSearchRequest()`~
12. Add `Toolbox::getNormalizedItemtype()`

## 1. Add `FilterableInterface` interface to allow any `CommonDBTM` to define search engine based filters

Implementing `FilterableInterface` will allow any `CommonDBTM` object to define some arbitrary filters.
This is done by adding a new "Filter" tab automatically to any class `CommonDBTM` implementing `FilterableInterface`:

![image](https://user-images.githubusercontent.com/42734840/228249931-96fbb415-f4b0-4a7d-8f9f-c3c037d13043.png)

Note: items need to redefine `defineTabs()` correctly (calling `parent::defineTabs()` at some point) for the tab to be automatically added.

The goal is that it should be as easy as possible to add a filter tab to any class (just implement `Filterable` and you are done).

You only need two methods to implements `Filterable`:

1\) `public function getItemtypeToFilter(): string;`

Indicate the itemtype to filter.
For example, in the screenshot above I am on the "New followup" notification and I want to filter the possibles tickets that may trigger this notifications, so the target itemtype here is `Ticket`.

2\) `public function getItemtypeField(): ?string;`

If the itemtype come from a dynamic data (e.g. a database field indicating the itemtype, which is the case for my notification above as some other notification may be targeting different itemtypes), then we need to store the information in this function.
That will allow the `CommonDBTM` parent class to clean up any filter if it detect an itemtype change (i.e. a filter defined for tickets wont be valid for any other itemtype and must be deleted).

## 2. Add `Item_Filter` class to store and interact with filters

The filters data are stored in the `Item_Filter` class, which also contain an helper method to check if an item match the given filter: 
```php
public static function itemMatchFilter(
    CommonDBTM $item,
    CommonDBTM&Filterable $filterable
): bool {
```

Going back to the notification example, it will be used when a notification event is raised.
If the item doesn't match the filter criteria then the notification will not processed.

## 3. Support filters for notifications targets by implementing `Item_Filter` and `Filterable` logic to Notifications

This is the main feature requested in the internal ref; filtering notification targets.

![image](https://user-images.githubusercontent.com/42734840/228249931-96fbb415-f4b0-4a7d-8f9f-c3c037d13043.png)

Here, the `Add followup` notification will only be sent for tickets associated to the `C1` computer.

## 4. Add `hide_controls` parameter to the search engine

When enabled, the seach results header will be omitted:
![image](https://user-images.githubusercontent.com/42734840/228255813-16a5decd-44e7-4ee6-8ca9-cd4eb1a9f05c.png)

This was needed to display the filter results, we don't need any user interaction here.

## 5. Add `extra_actions_templates` parameter to the search engine

Allow to pass extra twig template(s) to the search engine to add extra actions buttons.
This allow us to add two extra buttons to handle our filter data:

![image](https://user-images.githubusercontent.com/42734840/228256723-2926238d-1785-457e-9f7f-5eb97ba07f0b.png)

## 6. Add `push_history` parameter to the search engine

May be disabled to prevent the search engine to edit the current URL when submitting a search.

![image](https://user-images.githubusercontent.com/42734840/228257477-0c37e9a5-763d-4601-be69-10a231197135.png)

Needed as we don't want the URL above to be rewritten as a tickets search query when previewing a filter.

## 7. Fix display parameters being lost on search submit

The new search parameters affecting the results area such as `hide_controls` and the existing parameter `showmassiveactions` would only be taken into account when displaying the page.
Any search submitted after would not take into account these parameters.

This was fixed by adding them to the search's form and handling them accordingly.

## ~8. Fix js toasts inconsistencies~

Moved to https://github.com/glpi-project/glpi/pull/14520.

## 9. Fix metacriteria search

Metacriteria display was broken in main:

![image](https://user-images.githubusercontent.com/42734840/228259404-116ff2bf-95e9-49ab-90a3-f1960db66647.png)

The `p` variable wasn't passed correctly to the template.

## ~10. Fix `getItemTypeForTable()` namespace handling~

Moved to https://github.com/glpi-project/glpi/pull/14566.

## ~11. Add `CommonDBTM::getDefaultSearchRequest()`~

Reverted.

## 12. Add `Toolbox::getNormalizedItemtype()`

Moved some repeated code in a specific method.
Didn't end up using the method in the changes related to the new feature but kept it anyway.

| Q             | A
| ------------- | ---
| Bug fix?      | no (but contains some small fixes)
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !26075

cedric-anne commented on ajax/items_filter.php:
```diff
@@ -0,0 +1,128 @@
+        // Default values for this endpoint
+        $itemtype = $_POST['item_itemtype'] ?? false; // Note: "item_" prefix because the search engine already use the itemtype key
+        $items_id = $_POST['item_items_id'] ?? false;
+        $search_criteria = $_UPOST['criteria'] ?? []; // Note: criteria may be missing in a valid form + must be UPOST (see Item_Filter::saveFilter)
```
```suggestion
        $search_criteria = Sanitizer::unsanitize($_POST['criteria'] ?? []); // Note: criteria may be missing in a valid form + must be UPOST (see Item_Filter::saveFilter)
```

AdrienClairembault replied:
Done 06c0314c96.

cedric-anne commented on src/Search/Item_Filter.php:
```diff
@@ -0,0 +1,276 @@
+     *
+     * @return bool
+     */
+    public static function deleteFilter(CommonDBTM&Filterable $item): bool
+    {
+        $filter = self::getForItem($item);
+
+        // No filter, nothing to be done
+        if (!$filter) {
+            return true;
+        }
+
+        // Delete existing filter
+        return $filter->delete(['id' => $filter->fields['id']]);
+    }
```
Same here, could be moved to a `FilterableTrait` (`$this->deleteFilter()`).

cedric-anne commented on src/Search/Item_Filter.php:
```diff
@@ -0,0 +1,276 @@
+     *
+     * @return bool
+     */
+    public static function saveFilter(
```
Same here.

cedric-anne commented on templates/components/search/display_data.html.twig:
```diff
@@ -36,27 +36,31 @@
    <div class="ajax-container search-display-data">
 {% endif %}
+{% set push_history = push_history is defined and push_history %}
+{% set readonly = readonly is defined and readonly %}
```
Maybe `hide_controls` would be more appropriate than `readonly`.

AdrienClairembault replied:
Done 46ba760ad6.

cedric-anne commented on src/Search/Input/QueryBuilder.php:
```diff
@@ -473,6 +472,7 @@ public static function displayMetaCriteria($request = [])
             'num'         => $num,
             'itemtype'    => $itemtype,
             'linked'      => $linked,
+            'p'           => $p,
```
Ignore this.

AdrienClairembault commented on install/migrations/update_10.0.x_to_10.1.0/item_filter.php:
```diff
@@ -0,0 +1,59 @@
+$table = "glpi_searches_items_filters";
+if (!$DB->tableExists($table)) {
+    $query = "CREATE TABLE `$table` (
+        `id` int unsigned NOT NULL AUTO_INCREMENT,
```
Done 1c99fcb118.

AdrienClairembault commented on js/glpi_dialog.js:
```diff
@@ -341,16 +341,16 @@ const glpi_toast = (title, message, css_class, options = {}) => {
-    const animation_classes = options.animated ? `animate_animated ${options.animation} ${options.animation_extra_classes}` : '';
+    const animation_classes = options.animated ? `animate__animated ${options.animation} ${options.animation_extra_classes}` : '';
     let location = CFG_GLPI.toast_location || 'bottom-right';
     const valid_locations = ['top-left', 'top-right', 'bottom-left', 'bottom-right'];
     // If location is not valid, change it to bottom-right
     if (!valid_locations.includes(location)) {
         location = 'bottom-right';
     }
     const html = `<div class='toast-container ${location} p-3 messages_after_redirect'>
-      <div id='toast_js_${toast_id}' class='toast ${css_class} ${animation_classes}' role='alert' aria-live='assertive' aria-atomic='true'>
-         <div class='toast-header'>
+      <div id='toast_js_${toast_id}' class='toast border-0 ${animation_classes}' role='alert' aria-live='assertive' aria-atomic='true'>
+         <div class='toast-header ${css_class}'>
```
10.0.8 fix -> https://github.com/glpi-project/glpi/pull/14520

AdrienClairembault replied:
Toast changes reverted from this PR -> 1aa8af4bee

AdrienClairembault commented on ajax/items_filter.php:
```diff
@@ -0,0 +1,128 @@
+Session::checkLoginUser();
+
+// Read endpoint
+$action = $_POST['action'] ?? false;
```
Done f77d593be9.

AdrienClairembault commented on src/CommonDBTM.php:
```diff
@@ -1962,6 +1964,15 @@ public function post_updateItem($history = 1)
+        // Clear filter on itemtype change
+        if (
+            $this instanceof Filterable
+            && !is_null($this->getItemtypeField())
```
Done 0cfef72209.

AdrienClairembault commented on src/CommonDBTM.php:
```diff
@@ -6571,4 +6582,23 @@ public function isGlobal(): bool
+            'criteria' => [
+                [
+                    'link'       => 'and',
+                    'field'      => 'view',
```
I see that the `view` criteria can indeed be disabled.

I feel like there is a conflict between `getDefaultCriteria` and `getDefaultSearchRequest`.
Only one should exist (probably `getDefaultSearchRequest` as it can be overidden by children classes).

Should I delete `getDefaultSearchRequest` and use `getDefaultCriteria` everywhere instead (with the correct `view` code taken from `getDefaultSearchRequest`) ?

cedric-anne replied:
Most of plugins that are defining the `getDefaultSearchRequest` are just passing `sort`+`order` params, but no `criteria`.

If you define a `CommonDBTM::getDefaultSearchRequest()`, plugins will have to be updated to do something like:
```php
// PluginFormcreatorFormAnswer

    static function getDefaultSearchRequest()
    {
        return array_merge(
            parent::getDefaultSearchRequest(),
            [
                'sort'   => 6,
                'order' => 'DESC'
            ]
        );
    }
```

Sounds an acceptable change for me, but requires a changelog entry, and will probably be a pain for you to backport plugin patches on 10.0 version.

AdrienClairembault replied:
I guess I'll keep it simple for this PR and just fix `getDefaultSearchRequest`.

AdrienClairembault replied:
Done 2a925cd1b0.

AdrienClairembault commented on src/DbUtils.php:
```diff
@@ -264,9 +264,25 @@ public function getItemTypeForTable($table)
-
             $base_itemtype = $this->fixItemtypeCase($prefix . $table);
 
+            // Tables with namespace name
+            $namespace = "Glpi";
+            $base_itemtype_parts = explode("_", $base_itemtype);
+            foreach ($base_itemtype_parts as $base_itemtype_part) {
+                // Build namespace
+                $namespace .= "\\$base_itemtype_part";
+
+                // Rebuild class name from remaining part and current namespace level
+                $level = substr_count($namespace, "\\");
+                $remainings_parts = array_slice($base_itemtype_parts, $level);
+                $itemtype = "$namespace\\" . implode("_", $remainings_parts);
+
+                if (class_exists($itemtype)) {
+                    return $itemtype;
+                }
+            }
```
It didn't worked for the `glpi_searches_items_filters` table.

cedric-anne replied:
It is probably due to usage of both namespace and `_` in classname. Anyway, `XXX_YYY` is meant to be used for `CommonDBRelation`, with both `XXX` and `YYY` corresponding to an existing itemtype. Your class inherits `CommonDBChild`, you should name it `Glpi\Search\ItemFilter`.

AdrienClairembault replied:
Setting `CommonDBChild` was a mistake, it should indeed be a `CommonDBRelation` (like others Item_XXXX classes) -> I'll change that.

I think the issue is that `getItemTypeForTable` doesn't handle correctly `CommonDBRelation` + namespace format.

cedric-anne replied:
Could you please make a dedicated PR for this issue, on 10.0/bugfixes branch, and add corresponding cases in data providers of existing tests?

AdrienClairembault commented on src/Search/Item_Filter.php:
```diff
@@ -0,0 +1,276 @@
+     *
+     * @return bool
+     */
+    public static function itemMatchFilter(
```
Looking into it

AdrienClairembault replied:
Done fc6b91eba8.

AdrienClairembault commented:
Note: tests will fail until I can rebase on https://github.com/glpi-project/glpi/pull/14566.

AdrienClairembault commented:
UX tweaks from the lastest commit, suggested by @orthagh:

#### 1\) Info header

![image](https://user-images.githubusercontent.com/42734840/233361120-4623ee74-2084-416c-8896-ff364e70a3d1.png)

Since each item can choose what it will do with the defined filter, I've added an info area at the start of tab.
Item implementing `Filterable` must define the informations messages to display.

I've used boostrap "callouts" for the visual, which is not an official component but something that bootstrap use in their own documentation, for example:

![image](https://user-images.githubusercontent.com/42734840/233361567-e33a23f6-e7b7-4c9b-86e0-61cef8048724.png)

I think its a nice way to highlight some extra informations.

#### 2\) Do not preview the filter results on reload

When loading the tab we no longer preview the results:

![image](https://user-images.githubusercontent.com/42734840/233361959-f7103432-7a53-42cb-8188-04e88a9cae50.png)

This help emphasize the search criteria themselves instead of the search results (we are not selecting a set of items, we are defining a set of criteria).

#### 3\) Indicate clearly that there is no filter defined.

When looking at an item with no defined filter, seeing the default criteria might trick the user into thinking that they are already applied to the item.
There is still some hint that indicate that the filter is not applied (no delete button, count number in the tab, ...) but it might sill be ambiguous for the users.

To improve this, I've added a initialization step that cleary show that no filter exists:

![image](https://user-images.githubusercontent.com/42734840/233361922-b97ac7e6-a33a-4bc1-895a-47155193e812.png)

This is done using tabler "blank page" template (https://preview.tabler.io/blank.html) with some tweaks to match our layout/flow.

AdrienClairembault commented:
With that last UX change done it should be ready for final review / merge :)
Tests will still be failing until I can rebase on `main` with the `getItemtypeForTable` fix included (need a merge of 10.bf into main first).

AdrienClairembault commented:
Another small UI change suggested by @orthagh:

![image](https://user-images.githubusercontent.com/42734840/233366153-4bed78bd-f285-428a-bb99-c626cdba2607.png)

cedric-anne commented on css/includes/components/_callout.scss:
```diff
@@ -0,0 +1,83 @@
+ * Quick and easy way to make some text pop off
+ */
+
+.bd-callout {
```
What this `bd` stands for?

AdrienClairembault replied:
I would guess "bootstrap documentation".
Renamed to `callout` in 2ccaa7440e.

cedric-anne commented on templates/components/search/item_filter.html.twig:
```diff
@@ -0,0 +1,81 @@
+        <div id="manage_filter_ux" class="{{ filter_enabled ? "" : "d-none" }}">
+
+            {# Display search UX #}
+            {{ call('Glpi\\Search\\Input\\QueryBuilder::showGenericSearch', [itemtype, params]) }}
```
This method is outputing HTML directly and does not return a printable value, so you should not use `{{ ... }}` (it is made to print something).
```suggestion
            {% do call('Glpi\\Search\\Input\\QueryBuilder::showGenericSearch', [itemtype, params]) %}
```

cedric-anne commented on templates/components/search/item_filter.html.twig:
```diff
@@ -0,0 +1,81 @@
+            {# Display search results, hidden until a search is executed #}
+            <div id="item_filter_preview" class="d-none">
+                <h2>{{ __("Preview") }}</h2>
+                {{ call('Glpi\\Search\\SearchEngine::showOutput', [itemtype, params]) }}
```
```suggestion
                {% do call('Glpi\\Search\\SearchEngine::showOutput', [itemtype, params]) %}
```

cedric-anne commented on src/Search/Item_Filter.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     * @return null|self Null if no filter are defined for the given item
+     */
+    public static function getForItem(CommonDBTM&FilterableInterface $item): ?self
+    {
+        $filter = new self();
+        $filter_exist = $filter->getFromDBByCrit([
+            'itemtype' => $item->getType(),
+            'items_id' => $item->getID(),
+        ]);
+
+        return $filter_exist ? $filter : null;
+    }
```
This method is only used in `FilterableTrait`. It could be moved there, made private and non static.

AdrienClairembault replied:
Its also used in `Item_Filter`: 

![image](https://user-images.githubusercontent.com/42734840/233398933-d68de3f8-6ee5-48a2-8f67-31bae0676abe.png)

![image](https://user-images.githubusercontent.com/42734840/233399208-f6f6d686-479b-4725-8dc9-3c2c750663af.png)

I guess it could still be moved into the trait and called from $item (but obviously as a public method not private) but I'm not too sure as its not really related to the filter feature, it's just some kind of CommonDBChild boilerplate.

cedric-anne reviewed (commented):
Seems OK.

AdrienClairembault commented:
After seconds thoughts, the `Item_Filter` name is missleading as the class is a `CommonDBChild` not a `CommonDBRelation`.

I've renamed the class to `CriteriaFilter` to avoid confusion.
